spotify_dl/youtube.py

```diff
@@ -1,118 +1,270 @@
-                            type=3, desc=u'Cover',
-                            data=resp.read()
-                        )
-                song_file.save()
-            else:
-                print('File {} already exists, we do not overwrite it '.format(mp3filename))
+                    print(f"Failed to download {name}, make sure yt_dlp is up to date")
+            if not kwargs["skip_mp3"]:
+                set_tags(temp, file_path, kwargs)
+
+
+def multicore_find_and_download_songs(kwargs):
+    """
+    function handles divinding songs to be downloaded among the specified number of CPU's
+    extra songs are shared among the CPU's
+    each cpu then handles its own batch through the multihandler fn
+    """
+    reference_file = kwargs["reference_file"]
+    lines = []
+    with open(reference_file, "r", encoding="utf-8") as file:
+        for line in file:
+            lines.append(line)
+    cpu_count = kwargs["multi_core"]
+    number_of_songs = len(lines)
+    songs_per_cpu = number_of_songs // cpu_count
+    extra_songs = number_of_songs % cpu_count
+
+    cpu_count_list = []
+    for cpu in range(cpu_count):
+        songs = songs_per_cpu
+        if cpu < extra_songs:
+            songs = songs + 1
+        cpu_count_list.append(songs)
+
+    index = 0
+    file_segments = []
+    for cpu in cpu_count_list:
+        right = cpu + index
+        segment = lines[index:right]
+        index = index + cpu
+        file_segments.append(segment)
+
+    processes = []
+    segment_index = 0
+    for segment in file_segments:
+        p = multiprocessing.Process(
+            target=multicore_handler, args=(segment_index, segment, kwargs.copy())
+        )
+        processes.append(p)
+        segment_index += 1
+
+    for p in processes:
+        p.start()
+    for p in processes:
+        p.join()
+
+
+def multicore_handler(segment_index, segment, kwargs):
+    """
+    function to handle each unique processor spawned download job
+    :param segment_index: to be used for naming the reference file to be used for processor's download batch
+    :param segment: list of songs to be downloaded using spawning processor
+    """
+    reference_filename = f"{segment_index}.txt"
+    with open(reference_filename, "w+", encoding="utf-8") as file_out:
+        for line in segment:
+            file_out.write(line)
+
+    kwargs["reference_file"] = reference_filename
+    find_and_download_songs(kwargs)
+
+    if os.path.exists(reference_filename):
+        os.remove(reference_filename)
+
+
+def download_songs(**kwargs):
+    """
+    Downloads songs from the YouTube URL passed to either current directory or download_directory, as it is passed.  [made small typo change]
+    :param kwargs: keyword arguments to be passed on between functions when downloading
+    """
+    for url in kwargs["songs"]["urls"]:
+        log.debug("Downloading to %s", url["save_path"])
+    reference_file = "All Songs For This Download.txt"
```

Comment:
spaces for file names causes issues across platforms, please keep this to short name

Reply:
made the necessary changes

Reply:
ok corrected everything i can . hope i did not leave out a thing this time. for the deep source analysis. the general exception being caught wasnt introduced by I. if theres anyway i can rectify it, id be glad to know